interp/src/flatten/structures/environment/wave.rs

```diff
@@ -0,0 +1,220 @@
+use crate::flatten::flat_ir::prelude::*;
+use crate::flatten::structures::context::Context;
+use crate::flatten::structures::environment::{Environment, PortMap};
+use crate::flatten::structures::index_trait::IndexRef;
+use crate::flatten::structures::indexed_map::AuxillaryMap;
+use baa::BitVecOps;
+use fst_writer::*;
+
+#[derive(Debug, thiserror::Error)]
+pub enum WaveError {
+    #[error("FST write operation failed.")]
+    Fst(#[from] fst_writer::FstWriteError),
+}
+
+pub type Result<T> = std::result::Result<T, WaveError>;
+
+pub struct WaveWriter {
+    writer: Option<FstBodyWriter<std::io::BufWriter<std::fs::File>>>,
+    port_map: PortToSignalMap,
+}
+
+impl WaveWriter {
+    pub fn open<C: AsRef<Context> + Clone>(
+        file_path: &std::path::PathBuf,
+        env: &Environment<C>,
+    ) -> Result<Self> {
+        let info = FstInfo {
+            start_time: 0,
+            timescale_exponent: 0,
+            version: "Cider 2".to_string(),
+            date: "today".to_string(),
+            file_type: FstFileType::Verilog,
+        };
+        let mut writer = open_fst(file_path, &info)?;
+        let port_map = declare_signals(&mut writer, env)?;
+        let writer = writer.finish()?;
+        Ok(Self {
+            writer: Some(writer),
+            port_map,
+        })
+    }
+
+    pub fn write_values(&mut self, time: u64, values: &PortMap) -> Result<()> {
+        let writer = self.writer.as_mut().unwrap();
+        writer.time_change(time)?;
+        for (port_id, maybe_signal_id) in self.port_map.iter() {
+            if let Some(signal_id) = maybe_signal_id {
+                match values[port_id].val() {
+                    None => {
+                        writer.signal_change(*signal_id, "x".as_bytes())?;
+                    }
+                    Some(value) => {
+                        writer.signal_change(
+                            *signal_id,
+                            value.to_bit_str().as_bytes(),
+                        )?;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Drop for WaveWriter {
+    fn drop(&mut self) {
+        let writer = std::mem::take(&mut self.writer);
+        if let Some(writer) = writer {
+            writer.finish().unwrap();
+        }
+    }
+}
+
+enum Todo {
+    /// instance name, instance id
+    OpenScope(String, GlobalCellIdx),
+    CloseScope,
+}
+
+type PortToSignalMap = AuxillaryMap<GlobalPortIdx, Option<FstSignalId>>;
+
+fn declare_signals<
+    C: AsRef<Context> + Clone,
+    W: std::io::Write + std::io::Seek,
+>(
+    out: &mut FstHeaderWriter<W>,
+    env: &Environment<C>,
+) -> Result<PortToSignalMap> {
+    let mut port_map: PortToSignalMap = PortToSignalMap::new();
+    let root_idx = GlobalCellIdx::new(0);
+    let root_comp = &env.ctx.as_ref().secondary[env.ctx().entry_point];
+    let root_name = env.ctx().lookup_name(root_comp.name).clone();
+    let mut todo = vec![Todo::OpenScope(root_name, root_idx)];
```

Comment:
Can also do this, if desired. They're equivalent
```suggestion
    let root_name = root_comp.name.lookup_name(env.ctx().as_ref()).clone();
```